platforms/core-configuration/model-core/src/main/java/org/gradle/internal/properties/bean/DefaultPropertyWalker.java

```diff
@@ -127,7 +138,13 @@ public TaskDependencyContainer getTaskDependencies() {
         public void maybeFinalizeValue() {
             if (isConfigurable()) {
                 Object value = cachedInvoker.get();
-                ((HasConfigurableValueInternal) value).implicitFinalizeValue();
+                if (isUpgradedProperty) {
+                    // Upgraded properties should not be finalized to simplify migration.
+                    // This behaviour should be removed with Gradle 10.
+                    ((HasConfigurableValueInternal) value).warnOnUpgradedPropertyChanges();
+                } else {
+                    ((HasConfigurableValueInternal) value).implicitFinalizeValue();
+                }
```

Comment:
:x: I would do this differently. Instead of calling a different method here, I would implicitly finalize the value just as well. However, I would also call something like `HasConfigurableValueInternal.markAsUpgradedPropertyValue()` here:

```java
public CachedPropertyValue(Supplier<Object> supplier, Class<?> declaredType, boolean isUpgradedProperty) {
  this.declaredType = declaredType;
  this.cachedInvoker = Suppliers.memoize(() -> {
    Object value = DeprecationLogger.whileDisabled(supplier::get));
    if (isUpgradedProperty && isConfigurable()) {
      ((HasConfigurableValueInternal) value).markAsUpgradedProperty();
    }
    return value;
  }
}

I think if we do it like this, there's fewer modifications needed that are easier to remove in 10.

If you think about it on a conceptual level, it also makes sense: instead of introducing a new finalization mechanism, we are just altering what the existing implicit finalization means, which is a less impactful change.
```

Reply:
Ok, sure why not. I think changes needed are almost the same in the end, but we can go also with this approach